### PR TITLE
fix: generate peer-id on launch

### LINKED ISSUE #5232: generate torrent peer-ids on launch; do not periodically recycle
### What is the issue?

Other month I merely get a ban on my favorite torrent tracker. The reason told by its admin, was the following: during all my part in beta testing process, since October, 25 2022 up to February, 8 2023, my Transmission client has reported more than 700 (!) different ID's (looking like peer_id=-TR4010-XXXXXX, where XXXXXX changed many times a day).  That lead to "Could not insert peer details; Error: Duplicate entry '862-782ebf57' for key 'PRIMARY'" error and violating tracker rules as a consequence.
Since beta testing will go on, may respected developers improve something to avoid such an inconvenient situation?
Thank you very much.

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0 b1, b2, b3

### PR DESCRIPTION
**Pull request by ckerr**
Fixes #5232.

Notes: Updated Peer ID generation to happen once per session.

### CODE REVIEW

**Review by Coeur — APPROVED** [review]

**Comment by Tockman** [comment]
I must specify that my issue is about changing first, fixed part of peer_id (**TR4010**-XXXXXX), not the random part (TR4010-**XXXXXX**)!

**Comment by ckerr** [comment]
@Tockman I'm confused by what you mean then? That section doesn't change except when there's a version bump.

Could you be more descriptive about what the problem actually is? `"Could not insert peer details; Error: Duplicate entry '862-782ebf57' for key 'PRIMARY'` is not super helpful.

**Comment by Tockman** [comment]
I'm right now trying to get log details from the tracker's admin. Then my talk will become more specific.

**Comment by Tockman** [comment]
Here's a fragment of my clients' peer-id history. Unfortunately, there's not any full log of first 8 bytes (constant part) collection, just 4-bytes hash, but that's all we (they) have.
[Log.pdf](https://github.com/transmission/transmission/files/10996335/Log.pdf)

**Comment by Tockman** [comment]
The problem actually is my client peer-id during beta period changed more than 700 times, and changed exactly in unchangeable part of peer_id (**TR4010**-XXXXXX) that contains client version. That lead to one peers table's overflow related to limitation of GROUP_CONCAT in mysql, that is 1024 by default.

**Comment by ckerr** [comment]
I'm pretty sure the `-TR4010-` fragment did not change 700 times. It changed once per release, exactly as it should. When I look in the third column of that pdf, I only see the same 5 or so versions repeated again and again.

I don't understand what the complaint is nor what change is being suggested. It looks to me like Transmission behaved exactly as it's supposed to?

**Comment by ckerr** [comment]
@Tockman by any chance is this a private tracker and you have a torrent for that site that's incorrectly flagged as public?

**Comment by Tockman** [comment]
> @Tockman by any chance is this a private tracker and you have a torrent for that site that's incorrectly flagged as public?

No any chance. They are marked as private only. See the picture: 3-rd line means "Protection: Private torrent, users' discovery aside the tracker is disabled".

Снимок экрана 2023-03-17 в 11 05 38

**Comment by Tockman** [comment]
Yes, every one totally. Because in the moment of downloading torrent file is formed with your own passkey. There are no chances to get a public torrent - that's Graal of closed trackers' business. At least, at the tracker we are talking about.

**Comment by Tockman** [comment]
Well, there is a hypothesis - a speculation only - that betas could assign a unique peer-id to every task, downloading or uploading, changing it every time when TTL is terminated. This might be an explanation of such a strange behavior with 700+ peer-id's constant-part changes.

**Comment by ckerr** [comment]
You keep saying "700+". Every torrent gets its own unique peer-id. This is the same thing that [libtorrent does](https://github.com/arvidn/libtorrent/blob/8cfceaecafd3a7346b5f8d38f81ef7656101fd0b/src/torrent.cpp#L207) and [raskshasa does](https://github.com/rakshasa/libtorrent/blob/8492479e356933a34b9baf5de5a769b6863617b7/src/torrent/torrent.cc#L161).  How many torrents are you seeding on that tracker? Every one of them is going to have a unique peer id each time you restart Transmission. That is the same behavior you'll see in any torrent client. The first 8 characters of that peer id are going to be [`-TR4010-`](https://github.com/transmission/transmission/blob/280ace12f8a2b99fdf144011cd4f6fed8be876fe/CMakeLists.txt#L116) or whatever based on version number. The rest of the peer id is randomly generated. 

If the admin's complaint is that there were more than 1024 unique peer-ids at once, then what happens when someone is seeding more than 1024 torrents? Every torrent has a unique peer-id anyway, so....?

Overall I am just completely confused by this bug report. The discussion started with 

> -TR4010-XXXXXX, where XXXXXX changed many times a day

but then also

> I must specify that my issue is about changing first, fixed part of peer_id (**TR4010**-XXXXXX), not the random part (TR4010-**XXXXXX**)

and

> changed exactly in unchangeable part of peer_id (**TR4010**-XXXXXX) that contains client version

but then the pdf attachment only shows seven different versions of Transmission:

- `3.00`
- `3.00+`
- `4.0.0-beta.1`
- `4.0.0-beta.2`
- `4.0.0-beta.3`
- `4.0.0`
- `4.0.1`

So I am just completely confused about what the actual problem being reported is.

On the contradictory and incomplete info I've got so far there's nothing I can do with this issue. If you have more information or can clarify the complaint please open a new issue.

**Comment by ckerr** [comment]
> Well, there is a hypothesis - a speculation only - that betas could assign a unique peer-id to every task, downloading or uploading, changing it every time when TTL is terminated. This might be an explanation of such a strange behavior with 700+ peer-id's constant-part changes.

That doesn't seem to be borne out by the pdf log that you attached. There doesn't seem to be any pattern to the changes; it looks more like you were restarting the app every few days.

**Comment by Tockman** [comment]
The sad thing is my complaint built on fragmentary and incomplete information gotten from the tracker's admin. Well, I'll try to describe the situation once more.

1. March, 13 - I've found out that all my tasks at one of private trackers ceased to download or upload although TM shows all of them are connected to tracker announcer and report a number of seeds and leeches.
2. I apply to admin and ask for some help. The first hint I got is to check status of every task in the client. Funny.
3. Two days later I was told by admin about sone proprietary table of users' torrent clients that collects peer-id's but has a limitation of 1028 bytes for single user. Also I was told that in the run of last 3 months my beta versions of TM have sent 700+ different peer-id's, that resulted in table's overflow and automatic blocking of my side. I open my issue here.
4. You, Charles, make an assumption that problem is related to random part of peer-id, and I don't challenge it at first, because it seemed quite logical and obvious. But later I got a statement from admin that we are talking exactly about **constant** parts of peer-id (version related), not random ones. And, by the way, I was the single user at the tracker to have this phenomena (maybe, just because I was the single one with Transmission Beta on macOS).
5. I ask admin about any kind of proof: fragment of log, or at least an example of changed peer-id part. All I've got is a piece of table you've got before and mentioned above, with 4-byte hash and absolutely no specifics.
6. Any clarification of the purpose of mentioned table, and why it's so critical that its overflow interrupts user's communication with tracker, and may, or have its settings to be changed, - I didn't get.
7. I'm truly sorry that, trying to be fast, I reported there about wrong peer-id part at first, and unwittingly mislead you.
8. This fragment of table contents lines concerning not only betas, but 3.0 and 3.0+, and also stable 4.0 and 4.01, to make quantity of different peer-id reports comparable.
9. Sure, I hoped to get much more information on this. But unfortunately, I guess, that's all we have or can have now over that incident.

**Comment by ckerr** [comment]
> You, Charles, made an assumption that problem is related to random part of peer-id, and I didn't challenge it at first, because it seemed quite logical and obvious. 

My "assumption" was from [the bug report you filed](https://github.com/transmission/transmission/issues/5232#issue-1626352543):

> my Transmission client has reported more than 700 (!) different ID's (looking like peer_id=-TR4010-XXXXXX, where XXXXXX changed many times a day)."

**Comment by ckerr** [comment]
> 4. You, Charles, made an assumption that problem is related to random part of peer-id, and I didn't challenge it at first, because it seemed quite logical and obvious. 

That's not factually accurate. The claim that this came from the random part of the peer-id was made [by you in your description in the bug report](https://github.com/transmission/transmission/issues/5232#issue-1626352543):

> my Transmission client has reported more than 700 (!) different ID's (looking like peer_id=-TR4010-XXXXXX, where XXXXXX changed many times a day)."

**Comment by Tockman** [comment]
> > 7\. You, Charles, made an assumption that problem is related to random part of peer-id, and I didn't challenge it at first, because it seemed quite logical and obvious.
> 
> That's not factually accurate. The claim that this came from the random part of the peer-id was made [by you in your description in the bug report](https://github.com/transmission/transmission/issues/5232#issue-1626352543):
> 
> > my Transmission client has reported more than 700 (!) different ID's (looking like peer_id=-TR4010-XXXXXX, where XXXXXX changed many times a day)."

Well, my fault, mea culpa! Your assumption directly streamed from my mistake. Which one I, to be just fair, tried to correct as soon as it was possible.

**Comment by ckerr** [comment]
As is probably clear above, I'm a little frustrated with this ticket, and that's causing me to act less professionally than I should. Let me try again.

I'm frustrated because I'm being told "there is a problem" but not told what the problem actually _is_ -- 

It sounds like you're saying that Transmission announced to that tracker with over 700 different variations in the first 8 characters of a peer-id. This is not something that's backed up by evidence either in Transmission's code (where this is hardcoded), nor that pdf log (which only shows 7 versions), and is not consistent with your other discussion (talk about regenerating the peer id after a TTL expires, which never affects the first 8 characters).

So if it's not in the first 8 characters, then the complaint is that the random bytes are random. But several times you've said that's *not* the problem. So if it's not the first 8 characters, and it's not the remaining random characters... then what actually is the issue?

If I *had* to guess from the information given so far, it sounds like the tracker has a flaw that doesn't allow more than 1024 unique fingerprints from a given client at any given time. But that  seems like a stretch too, since that flaw it would prevent anyone from seeding more than 1024 torrents at once on that tracker.

tldr this bug report does not make sense. I need more information in order to even understand what issue is being reported and cannot proceed. If the admin has more information or wants to reach out to me directly at `${firstname}@${firstname}${lastname}.com` that would be fine.

**Comment by Tockman** [comment]
I'm in the dumbest position of all possible ones. I'm not a network pro, and all I write here is just, to say, a broadcast of some other men's statements and versions.

_It sounds like you're saying that Transmission announced to that tracker with over 700 different variations in the first 8 characters of a peer-id._
Yes! that's right completely. It's almost unbelievable from professional poiont of view, but that's a thing I've received from admin's.
_This is not something that's backed up by evidence either in Transmission's code (where this is hardcoded), nor that pdf log (which only shows 7 versions_
Alas, right.
_and is not consistent with your other discussion (talk about regenerating the peer id after a TTL expires, which never affects the first 8 characters)._
It was just a speculation - as far as I can understand now, quite unreal one, keeping in mind that changes were found (if any) in unchangeable zone of peer-id.
_So if it's not in the first 8 characters_
Wrong. It's exactly first 8 characters (leaving alone another 12 that are changed by default). Look, who will complain if the changeable part has changed, even 7 hundred times?! The great amusement of our admin was caused exactly by that fact! Along with that, he couldn't give an example of **changed 8 characters** - because 2 months have come and logs are erased.
_it sounds like the tracker has a flaw that doesn't allow more than 1024 unique peer ids from a given client at any given time_
Yes. And it looks like not that a flaw but kind of deliberate setting.
 _But that that flaw is so serious that it seems unlikely, since it would make it impossible to seed more than 1024 torrents at once on that tracker._
That's quite possible since many TM users noticed serious slow down when quantity of torrents in it reaches 1000. What's for me, maximum I had for that tracker was 370.
_If the admin has more information or wants to reach out to me directly_
I have a feeling that his aim is, on the contrary, to hide some details and leave all settings as they are. But I can be wrong, of course.

**Comment by Tockman** [comment]
> I need more information in order to even understand what issue is being reported and cannot proceed.

To call a spade a spade, it's a bug of specific tracker setting, not Transmission itself. But the single goal I wanted to achieve reporting this - to make possible normal uninterrupted work of TM even in such circumstances. It it is possible at all. Thank you for your time and patience, Charles!

**Comment by ckerr** [comment]
>> It sounds like you're saying that Transmission announced to that tracker with over 700 different variations in the first 8 characters of a peer-id.

> Yes! that's right completely. It's almost unbelievable from professional point of view, but that's a thing I've received from admin's.

"Almost" is not the word I would use. I only count seven versions in that pdf and they all map to known versions (e.g. betas, stable releases, or nightly builds).

If there are 700 variations of the first 8 characters please provide a sampling of them.

**Comment by ckerr** [comment]
I think there is some miscommunication going on here; what you're suggesting should be literally imposssible since those first eight characters are built into Transmission at compile time and cannot be changed without rebuilding or using a hex editor.

Can you please paste to me the exact conversation that the admin had with you? I would like to read their words for myself.

**Comment by Tockman** [comment]
OK, please:

Если бы вы внимательно посмотрели дамп, то увидели бы в нём:
Could not insert peer details; Error: Duplicate entry '862-782ebf57' for key 'PRIMARY'

Ваши беты почему-то представлялись по-разному, за примерно 100 дней было 700 (!) уникальных ID, что вызвало проблему. У большинства пользователей за всё время работы таких ID 1-2-3-4. В это же время другие пользователи использовали аналогичные беты, у них всё нормально. В дальнейшем такой ситуации быть не должно. 

Почему у остальных официальные бета-версии проблем не вызвали? Эти версии использовали много пользователей. У вас проблемы были только с бетами. C 3.00+, 4.0.0 и 4.0.1 такого не было. Если вы посмотрите свой дамп, там при обращении к анонсеру есть peer_id=-**TR4010**-XXXXXX, выделенная часть только у вас менялась 700 раз (здесь закодировано TR - Transmission, 4010 - версия 4.01). Ситуация ненормальная и это является нарушением.

**Comment by Tockman** [comment]
peer-id - 20 символов. Первые 8 фиксированные, 12 случайные. В первых 8 задается тип клиента и номер версии. Последние 12 можно менять с любой частотой.
Так вот, у вас менялись первые 8 символов, чего быть не может. Использование кривого клиента, как в данном случае, нарушение. Я уже писал, что у пользователей разных версий по несколько, у вас 700. Больше я повторяться не хочу, читайте внимательно что я писал.

**Comment by Tockman** [comment]
Никто вас не отключал. За счёт переполнения таблицы получилось то, что получилось. Мне пришлось потратить время, чтобы с этим разобраться.

Q.: что из себя представляет эта запись, в какую таблицу добавляется и как это добавление соотносится с объяснением проблемы?
A.: Это внутреннее устройство, долго объяснять. В итоге попадали на ограничение GROUP_CONCAT в mysql - по умолчанию 1024.

**Comment by Tockman** [comment]
As you can see, admin considers that it is impossible, just like you.  And the same time he assures it was!

**Comment by garoto** [comment]
Reading this PR and the issue its related to, it's obvious by now to me that if any of the still active transmission variants around (v2.9x, v3.xx or the newer v4.x.x) had a `peer_id` schizo-spam issue, operators from top-tier trackers would had take action by now in 2023. And, as luck would have it, nothing is happening regarding transmission being a bad client with quality-operated trackers, like it have always been.

**Comment by andreygursky** [comment]
> edit: the root cause of OP's bug report appears to not be a Transmission bug.

Being curious, how this could happen. Provided Transmission sets the version in the fixed part correctly (which seems obvious) and the tracker uses exactly these 8 bytes (no other complaints), how the last could be different so often? One possible explanation: memory corruption in Transmission, which looks like has been fixed in 4.0.0.

**Comment by ckerr** [comment]
NFI. There is something that doesn't add up and I think we either have incomplete or inaccurate info from somewhere. It's not productive to speculate where that's coming from, IMO.

Just for thinking out the hypothetical -- if it were some kind of memory corruption, we would see more feedback from more users. There is something unique going on here that doesn't add up.

**Comment by ckerr** [comment]
Also if peer ID was being corrupted due to memory corruption, the tracker wouldn't be able to recognize it as Transmission

**Comment by andreygursky** [comment]
> more feedback from more users

It looks like some very infrequent event (even for those "huge" 700 times it took several months), which one would only proving by running always Wireshark. Since on the other side just only one specific database design with a corresponding limitation made it visible, otherwise it went unnoticed.

> the tracker wouldn't be able to recognize it as Transmission

Before writing the initial message, I've been thinking about it also. Until I realized, that the tracker got it explicitly just from "User Agent", which Transmission sets using ```CURLOPT_USERAGENT``` to "Transmission/SHORT_VERSION_STRING". This fact made me pretty sure, the tracker might indeed got garbage in the fixed part.

**Comment by garoto** [comment]
My bad commenting yet again on this already commited PR, but perhaps consider reverting this commit then, since it was 100% concocted based on a very poorly documented issue. AFAIK, transmission-deamons as far as v2.94 are still behaving 'just fine' on a variety of private trackers.
